Pennsylvania County-Level Official Scraper

christopherturner opened the pull request:
Good afternoon!  I was hoping to have someone take a look at my code for the ported Pennsylvania scraper and offer feedback based on requirements, coding practices, or domain-specific knowledge.

At this point, the scraper includes all data from the `cmdc-tools` scraper, with the notable exception of ventilator data – the respective variables do not appear to have been added to `meta.covid_variables` yet.

I was specifically hoping that someone could take a look at my mappings of argcis columns to `meta.covid_variables`.  As this is my first scraper, I wanted to make sure that the mappings I've made are congruent with the definitions provided on the [Pennsylvania dashboard info page](https://www.health.pa.gov/topics/disease/coronavirus/Pages/Cases.aspx).

sglyon commented:
Awesome, thanks @christopherturner . 

I will look at this more closely tomorrow. At first glance I notice that the code is very clean with great style. We love that!

christopherturner commented on can_tools/scrapers/official/PA/pa_state.py:
```diff
@@ -0,0 +1,143 @@
+from typing import Dict
+
+
+class Pennsylvania(ArcGIS, DatasetBase):
```
I've refactored my code into two classes in the latest commit!

christopherturner commented on can_tools/scrapers/official/PA/pa_state.py:
```diff
@@ -0,0 +1,143 @@
+            ),
+        }
+
+        out = (
```
Thanks for the heads-up!  I've integrated use of _esri_ts_to_date() in my latest commit for the PennsylvaniaHospitals class.

christopherturner commented:
Additionally, the following is the reference that I used to determine the measurements and units that my code reflects.  On the [Pennsylvania Data](https://www.health.pa.gov/topics/disease/coronavirus/Pages/Cases.aspx) website, it's stated that:

> Total cases, negatives, and recovered represent unique individuals. Total PCR Tests can include numerous tests for the same individual. 
> Total case counts include confirmed and probable cases. 
> Negative case data only includes negative PCR tests. Negative case data does not include negative antibody tests.
> Individuals who have recovered is determined using a calculation, similar to what is being done by several other states. If a case has not been reported as a death, and it is more than 30 days past the date of their first positive test (or onset of symptoms) then an individual is considered recovered.

sglyon commented:
This is looking good, thanks @christopherturner 

I need to see why the tests are failing. I think, but don't know for sure yet, that the issue is your fork of this repo doesn't have permissions to push the docker container for postgres to this repo's set of docker containers :thinking: 

I'll take a look though

sglyon commented on can_tools/scrapers/official/PA/pa_state.py:
```diff
@@ -0,0 +1,131 @@
+            ),
+        }
+        out = (
+            df.melt(id_vars=["county"], value_vars=crename.keys())
```
Sorry, this changed from underneath you --  we renamed `county` to `location_name` to more closely match the database schema.

I'm already in the file, so I'll make the change myself!

sglyon commented:
Hey @christopherturner  I took another look and found a few things:

- The "county" column from the data source includes an entry for Pennsylvania, the whole state. I filter that out
- I needed to make a change to the DatasetBase class to compare county names and other strings using all lowercase letters -- there were some errors with captialization. Totally not in scope for your PR, but it fixed a test so I wanted to point out why I did it in case you look at my changes and wonder what I did
- I commented out the probable cases variable. That isn't a very precise variable and we found that it wasn't very useful so we stopped recording it.

There is one more change we need to figure out...

For the hospital scraper the `county` column also includes things like hospital regions. We need to come up with a solution for how to remove those and only keep data for counties. @cc7768 any ideas?

sglyon commented:
Note, the hospital regions include

```
Out[19]: 
array(['East Central HCC', 'HCC of Southwest PA', 'Keystone HCC',
       'North Central HCC', 'Northeast', 'Northcentral', 'Northeast HCC',
       'Northern Tier HCC', 'Northwest', 'Southcentral', 'Southeast HCC',
       'Southeast', 'Southwest'], dtype=object)
```

christopherturner commented:
On the Florida scraper, I noticed that some rows were dropped based on a query like this: `.query("location not in (12998, 12999)")`.  My most basic approach would be excluding those hospital regions, based on the above keys, using a query like this.  Looking at the ArcGIS table, I don't see any values where we could expect counties to be categorically different from hospital regions (it seems most everything is non-null in both cases).

sglyon commented:
@cc7768 reminded me today that  we actually have room in the schema for collecting data by hospital region.

I have a change in the works that will make it easier to add new locations like this, fix the test runner issues here, and make it easier to test locally. Should be in tomorrow. Stay tuned and I'll work with you to add the hospital region data!

sglyon commented:
Hey @christopherturner the updates to CI and database interactions are done now.

If we can merge in the main branch into this one we should be able to pick those changes up. Are you comfortable doing that merge here?

christopherturner commented:
Sounds great, @sglyon!  And definitely – I can pull the upstream and merge it into my branch/PR!

sglyon commented:
@christopherturner thanks for doing the merge.

I pushed one commit that added the Penn scrapers to the `__init__.py ` files and now we have a failing test because we haven't taught the database about the hospital regions.

@cc7768 do you think it is worthwhile to add these hospital regions to the database and collect data for them? Or should we follow @christopherturner suggestion to filter them out like we do non-counties in FL?

cc7768 commented:
Let's go ahead and filter them out for now and work to get this merged and live.

We can re-visit this (and other states with hospital regions) in the future

christopherturner commented:
Thanks for all of the feedback @sglyon and @cc7768!  In my latest commit, I've added the array of non-county regions and filtered out any rows that contain those values in the `location_name` column.

I've also moved the filter for `location_name != "Pennsylvania"` into that logic, rather than filtering it out in the return statement of `PennsylvaniaHospitals.pre_normalize()`.

sglyon commented:
Great work @christopherturner !
